Codechange: add and use the ability to store (lists of) structs in savegames

TrueBrain opened the pull request:
## Motivation / Problem

In a few places (stations, vehicles, towns, companies, ..) the SaveLoad code needs to handle either structs in structs, or lists of structs in structs.

This is all done manually, and not only hard to read, also easy to mess up.


## Description

Introduce `SL_STRUCT` and `SL_STRUCTLIST` to embed a struct in another struct.

For example, for Vehicles it used to have a special `SL` command to embed the common vehicle structure. This meant that SaveLoad needed to know about Vehicles.

Now, the common vehicle sub-struct is called `SlVehicleCommon` and to save it you can use:
```
SLEG_STRUCT(SlVehicleCommon)
```

This works with structs in structs in structs in structs in structs in ... just fine too.

NOTE: I made several commits to introduce `SL_STRUCT` in the rest of the code, as I am hoping that makes reviewing easier. 
NOTE 2: I left a lot of odd constructs in there, just to make review easier. The code really could use a cleanup, but .. I am trying not to do everything at once, also for my own sanity :D
NOTE 3: there should be no difference before or after this PR in respect to what is written/loaded from disk.

## Limitations

- I tested this with ~500 savegames, ranging from TTDp till 1.11.2, validating that they load and run. Nothing that wasn't broken before this PR is broken now. So I think there are no regressions, but with this many changes in the SaveLoad code, that cannot be guaranteed.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TrueBrain committed:
Codechange: remove the special station/vehicle code from SaveLoad

With the new SLEG_STRUCT it is much easier to embed a struct
in a struct, where the sub-struct has limitations on when it is
being used.
This makes both the code easier to read (less magic) and avoids
the SaveLoad needing to know all these things about Stations
and Vehicles.

rubidium42 commented on src/saveload/cheat_sl.cpp:
```diff
@@ -14,37 +14,60 @@
+	size_t count = SlGetFieldLength();
 	/* Cannot use lengthof because _cheats is of type Cheats, not Cheat */
-	if (count > sizeof(_cheats) / sizeof(Cheat)) SlErrorCorrupt("Too many cheat values");
+	if (count / 2 > sizeof(_cheats) / sizeof(Cheat)) SlErrorCorrupt("Too many cheat values");
```
Maybe you could use `count > lengthof(_cheats_desc)`? Then you could also remove the nasty dummies.

Having said that, this commit does not seem to be part/mentioned in this PR. Are you trying to cheat it in? ;)
Or did you want to consider using SLE_STRUCT for each of the Cheats?

TrueBrain replied:
This commit message making this change literal is called: `Codechange: rework CHTS-chunk save/load to be more like the others` :P I have to see if I have enough work to warrant another "mixed-fixes-for-stupid-code" PR .. but that is also why this PR is still a draft :)

Either way, this was existing code, I really did not parse yet what it said or tries to say. Just moved a line of code from one above to the one below.

What do you mean with "nasty dummies"? What "nasty" and what "dummies" is that referring to?

rubidium42 replied:
For the dummies, take a look at: https://github.com/OpenTTD/OpenTTD/blob/1e564b333fa2d7bde2f21d7a05bb3e81305185c5/src/cheat_type.h#L31 further on line 33 and 36. That, together with the comment around line 23 made them nasty to me.

rubidium42 commented on src/saveload/saveload.h:
```diff
@@ -399,6 +399,27 @@ struct ChunkHandler {
+/** A table of SaveLoad entries. */
+using SaveLoadTable = span<const struct SaveLoad>;
+
+class SaveLoadHandler {
```
I guess the draftiness of this PR means that the comments here are still pending / that you're waiting on glx's work, right?

TrueBrain replied:
it means a lot of things, but most that this is not the final form yet, and that I didn't do a pass myself on any coding guideliness :) I mostly created the PR so I could collect opinions on the way of doing stuff.

rubidium42 commented on src/saveload/town_sl.cpp:
```diff
@@ -113,7 +113,89 @@ void UpdateHousesAndTowns()
+public:
+	SaveLoadTable GetDescription()
+	{
+		/* Here due to private member vars. */
```
Is this still needed? I can't see `GetTileMatrixDesc` being made `friend` anywhere, so it seems to be not needed anymore.

rubidium42 commented on src/saveload/saveload.h:
```diff
@@ -692,11 +690,8 @@ struct SaveLoad {
-
-#define SLE_VEH_INCLUDE() {SL_VEH_INCLUDE, 0, 0, SL_MIN_VERSION, SL_MAX_VERSION, 0, [] (void *b, size_t) { return b; }, 0}
-#define SLE_ST_INCLUDE() {SL_ST_INCLUDE, 0, 0, SL_MIN_VERSION, SL_MAX_VERSION, 0, [] (void *b, size_t) { return b; }, 0}
+/** Only write byte during saving; never read it during loading. */
```
The comment is misleading as the byte is read, just separately using `SlReadByte` to ensure the right (sub)class is constructed. Now it reads like: "we write a byte to the save game, that we never need", and I start wondering. Why write a byte to the save game that is never read?

TrueBrain replied:
I am happy to read that you read it like that, as that is exactly what it does. And sure, we have an intended usage of it, but that is just that: how we intend someone to use it .. but nothing is enforcing that.

It really is a shitty mechanism that is very easy to misunderstand and fuck up. So I am more looking into how to remove it completely ;)

That said, I can add a line of doxygen comment to explain the intended use of it .. but at least you now understood correctly what it did :) The initial comment was truly misleading ;)

TrueBrain committed:
Codechange: ability to store structs and list of structs in savegames

The commits following this will use this new functionality.

Currently, a few places do this manually. This has as drawback that
the Save() and Load() code need to be in sync, and that any change
can result in (old) savegames no longer loading. In general, it is
annoying code to maintain.

By putting everything in a description table, and use that for
both Save() and Load(), it becomes easier to see what is going on,
and hopefully less likely for people to make mistakes.

TrueBrain committed:
Codechange: use SLE_STRUCT(LIST) for Town chunks

TrueBrain committed:
Codechange: use SLE_STRUCT(LIST) for Company chunks

TrueBrain committed:
Codechange: use SLE_STRUCT(LIST) for Station chunks

TrueBrain committed:
Codechange: use as much of STNN code for STNS as possible

There was a lot of code duplication for no real reason. Now with
SLEG_STRUCT support, we can just re-use the code, hopefully making
it easier for future-us to make changes to this, without breaking
everything for old games.

TrueBrain committed:
Codechange: use SLE_STRUCT(LIST) for GameScript chunks

TrueBrain committed:
Codechange: move Save/Load functions of same chunk next to each other

TrueBrain committed:
Codechange: use SLE_STRUCT(LIST) for Linkgraph chunks

rubidium42 commented on src/saveload/vehicle_sl.cpp:
```diff
@@ -812,66 +878,98 @@ SaveLoadTable GetVehicleDescription(VehicleType vt)
-	};
+	void Save(Vehicle *v) const override { this->GenericSaveLoad(v); }
+	void Load(Vehicle *v) const override { this->GenericSaveLoad(v); }
+	void FixPointers(Vehicle *v) const override { this->GenericSaveLoad(v); }
```
I'm seeing these three return over and over again, might it be worth it to have an intermediate "super" class with virtual GenericSaveLoad, or maybe even a templated intermediate super class that requires the implementing sub class to provide a GenericSaveLoad? The same is already done with description, and it means less duplicating of these three lines all over the save load code.

TrueBrain replied:
Later PRs mostly remove this; it is simply as a stop-gap. As I knew if I would to copy/paste the code and create identical functions I would get the complaints I have duplicated code :P This is the best I could do in the chain of PRs :)  Adding an abstraction now would just be a waste of my time, basically, as I would remove it 1 PR later :P I am lazy :D

But even after all this work there is plenty of room to refactor the code to have less code. But I left this out of this set of PRs, as reviewing this is already difficult enough :)

(just to confirm for myself, after #9322:
```
$ grep -R -n GenericSaveLoad ../src/ | wc -l
0
```
)

rubidium42 commented on src/saveload/saveload.h:
```diff
@@ -400,6 +400,68 @@ struct ChunkHandler {
+ * This handles a few common things for handlers, meaning the actual handler
+ * needs less code.
+ */
+template <class TOwn, class TObject>
```
Doxygen misses for these parameters.

rubidium42 replied:
TOwn -> TInstance or TImpl? Now it's a bit confusing w.r.t. Towns.

rubidium42 commented on src/saveload/station_sl.cpp:
```diff
@@ -374,89 +374,118 @@ static void Ptrs_STNS()
 };
 
-static OldPersistentStorage _old_st_persistent_storage;
+class SlStationCommon : public DefaultSaveLoadHandler<SlStationCommon, BaseStation> {
```
Maybe SlNormalStation as now it sort-of implies there might be a sub class?

rubidium42 commented on src/saveload/station_sl.cpp:
```diff
@@ -374,89 +374,118 @@ static void Ptrs_STNS()
-	      SLE_VAR(BaseStation, owner,                  SLE_UINT8),
-	      SLE_VAR(BaseStation, facilities,             SLE_UINT8),
-	      SLE_VAR(BaseStation, build_date,             SLE_INT32),
+class SlStationBase : public DefaultSaveLoadHandler<SlStationBase, BaseStation> {
```
I'd note in the comment for this one that SlWaypoint also makes use of this.
